cpp/src/parquet/arrow/arrow-reader-writer-test.cc

```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
   ASSERT_RAISES(Invalid, writer->WriteTable(*tbl, 1));
 }
 
+// ----------------------------------------------------------------------
+// Tests for directly reading DictionaryArray
+class TestArrowReadDictionary : public ::testing::TestWithParam<const char*> {
+ public:
+  void SetUp() override {
+    InitFromJSON(GetParam());
+    ASSERT_NO_FATAL_FAILURE(
+        WriteTableToBuffer(expected_dense_, expected_dense_->num_rows() / 2,
+                           default_arrow_writer_properties(), &buffer_));
+
+    properties_ = default_arrow_reader_properties();
+  }
+
+  void InitFromJSON(const char* jsonStr) {
+    auto dense_array = ::arrow::ArrayFromJSON(::arrow::utf8(), jsonStr);
+    expected_dense_ = MakeSimpleTable(dense_array, /*nullable=*/true);
+
+    ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+    const auto& string_array = static_cast<const ::arrow::StringArray&>(*dense_array);
+
+    for (int64_t i = 0; i < string_array.length(); ++i) {
+      if (string_array.IsNull(i)) {
+        ASSERT_OK(builder.AppendNull());
+      } else {
+        ASSERT_OK(builder.Append(string_array.GetView(i)));
+      }
+    }
+
+    std::shared_ptr<::arrow::Array> dict_array;
+    ASSERT_OK(builder.Finish(&dict_array));
+    expected_dict_ = MakeSimpleTable(dict_array, /*nullable=*/true);
+
+    // TODO(hatemhelal): Figure out if we can use the following to init the expected_dict_
+    // Currently fails due to DataType mismatch for indices array.
+    //    Datum out;
+    //    FunctionContext ctx(default_memory_pool());
+    //    ASSERT_OK(DictionaryEncode(&ctx, Datum(dense_array), &out));
+    //    expected_dict_ = MakeSimpleTable(out.make_array(), /*nullable=*/true);
```

Comment:
What is the error?

Reply:
Swapping in the cast implementation for setting up the `expected_dict_` table results in an a test assertion failure in `AssertTablesEqual`.  I used a debugger to work out that the trouble is the dictionary generated by the cast has indices with type `arrow::Int32Type` while the builder generated one has indices with type `arrow::Int8Type`.

I don't know if this is actually an artifact of the trivial test cases that I used?